ci/github-script/merge: various improvements

wolfgangwalther opened the pull request:
Some wording improvements, a TODO and preparation for both #451324 and #451362.

The nested checklists look like https://github.com/NixOS/nixpkgs/pull/451362#discussion_r2482795813.

## Things done

- [x] Tested locally again.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
ci/github-script/merge: improve wording

wolfgangwalther committed:
ci/github-script/merge: add TODO about second merge method

We have not observed this merge method being used in practice, yet. Not
in the new bot, not in the old bot. It seems like auto-merge works for
all cases.

wolfgangwalther committed:
ci/github-script/merge: support OR conditions

This supports AND on the first and OR on the second level, which is
needed for some follow up work like backports, approval based merges or
trusted maintainers.

MattSturgeon commented on ci/github-script/merge.js:
```diff
@@ -217,19 +218,20 @@ async function handleMerge({
-      `- :${canUseMergeBot ? 'white_check_mark' : 'x'}: ${comment.user.login} can use the merge bot.`,
-      `- :${isEligible ? 'white_check_mark' : 'x'}: ${comment.user.login} is eligible to merge changes to the touched packages.`,
+      `- :${isMaintainer ? 'white_check_mark' : 'x'}: ${comment.user.login} is a member of the [nixpkgs-maintainers](https://github.com/orgs/NixOS/teams/nixpkgs-maintainers) team.`,
+      `- :${isEligible ? 'white_check_mark' : 'x'}: ${comment.user.login} is eligible to merge changes to all touched packages.`,
```
Should this be more explicit?

```suggestion
      `- :${isEligible ? 'white_check_mark' : 'x'}: ${comment.user.login} is a maintainer of all touched packages.`,
```

Or is it intentionally vague to allow for future changes to eligibility?

wolfgangwalther replied:
I can't say whether they are a maintainer of all touched packages, because I am only testing for those in by-name. So once a non-by-name package is in the mix, the proposed change might not reflect reality anymore.

MattSturgeon replied:
So the alternative would be:
```suggestion
      `- :${isEligible ? 'white_check_mark' : 'x'}: ${comment.user.login} is a maintainer of all touched packages in `pkgs/by-name`.`,
```

Which is more explicit, but a little wordy.

wolfgangwalther replied:
The refactoring for https://github.com/NixOS/nixpkgs/pull/457652#discussion_r2484681445, also allowed me to only check this condition when all packages are in by-name, thus I went with the first proposal.

philiptaron reviewed (approved):
LGTM after CI is happy.

MattSturgeon reviewed (approved):
Nice!

(CI is complaining about treefmt)

<details><summary>treefmt diff</summary>
<p>

```diff
diff --git a/ci/github-script/merge.js b/ci/github-script/merge.js
index 14b1d8301..899a05b38 100644
--- a/ci/github-script/merge.js
+++ b/ci/github-script/merge.js
@@ -45,8 +45,8 @@ async function runChecklist({ github, context, pull_request, maintainers }) {
       'staging',
       'staging-next',
     ].includes(pull_request.base.ref),
-    'PR touches only packages in `pkgs/by-name/`.': files.every(({ filename }) =>
-      filename.startsWith('pkgs/by-name/'),
+    'PR touches only packages in `pkgs/by-name/`.': files.every(
+      ({ filename }) => filename.startsWith('pkgs/by-name/'),
     ),
```

</p>
</details>

emilazy commented on ci/github-script/merge.js:
```diff
@@ -45,19 +45,23 @@ async function runChecklist({ github, context, pull_request, maintainers }) {
+      'Authored by r-ryantm.': pull_request.user.login === 'r-ryantm',
+    },
+    [`PR has ${eligible.size} maintainer(s) eligible to merge changes to all touched packages.`]:
+      eligible.size > 0,
```
This is quite an unwieldy message for what seems like the least “relevant” criterion. Is this check strictly redundant to the other checks (i.e. it will never fail when the others pass), or is it actually checking something that isn’t covered by others?

MattSturgeon replied:
IIUC, the motivation is to express whether or not there are _any_ edible users, particularly in cases where the user in question is _not_ eligible.

I agree, this isn't really relevant when the user in question _is_ eligible.

I also suggested (in DMs) that it'd be better to list the users who are eligible, rather than simply stating "3 users can merge".

The issue here is we only have the maintainers' github _IDs_, not their usernames. We'd have to lookup each username separately via `GET /user/:id`, which eats into our ratelimit for little benefit.

Maybe that's not a big deal if we only do it when the current user isn't eligible, since that _should_ be fairly rare.

It's questionable whether its worth the complexity, though. Maybe that justifies a separate PR, assuming we want to implement it at all?

MattSturgeon replied:
> but this won't help for labeling

I think this user-facing comment and the internals of how we implement labeling are two separate issues, right?

I don't see why "we need it for labeling" should imply "we need to include it in the comment's checklist unconditionally".

emilazy replied:
My feeling is that this isn’t useful to list in the checklist if it will never be the cause of a failed merge.

wolfgangwalther replied:
I refactored the checklist stuff, so that this condition can now easily be checked for labeling only. It won't appear in the comments anymore.

wolfgangwalther committed:
ci/github-script/merge: improve caching of team members

This removes the need to `await` committers further down in the function
and allows re-using the cache for other teams later.

wolfgangwalther committed:
ci/github-script/merge: move API requests out of runChecklist

This makes runChecklist mostly a pure function (except for logging) to
allow calling it repeatedly later.

wolfgangwalther committed:
ci/github-script/merge: move all conditions into runChecklist

No special casing anymore, all conditions are in the same place. This
also has the benefit of hiding the "has maintainers eligible for merge"
condition from comments, because it is only really relevant for
labeling.

wolfgangwalther committed:
ci/github-script/merge: improve maintainer check

MattSturgeon commented on ci/github-script/merge.js:
```diff
@@ -98,15 +104,34 @@ async function handleMerge({
-    context,
-    pull_request,
-    maintainers,
+  function cache(team_slug) {
```
```suggestion
  function getTeamMembers(team_slug) {
```

MattSturgeon commented on ci/github-script/merge.js:
```diff
@@ -86,6 +88,10 @@ async function handleMergeComment({ github, body, node_id, reaction }) {
   )
 }
 
+// Caching the list of team members saves API requests when running the bot on the schedule and
```
Is it more accurate to say we're [memoizing](https://en.wikipedia.org/wiki/Memoization) the members, or fetching them lazily? We aren't writing to  a persistent cache.


```suggestion
// Memoizing team members saves API requests when running the bot on the schedule and
```

MattSturgeon reviewed (commented):
Looks good. A couple follow-up wording nits (not blocking):

wolfgangwalther commented on ci/github-script/merge.js:
```diff
@@ -45,19 +27,39 @@ async function runChecklist({ github, context, pull_request, maintainers }) {
+    'PR touches only packages in `pkgs/by-name/`.': allByName,
+    'PR is at least one of:': {
+      'Authored by a committer.': committers.has(pull_request.user.id),
+      'Created by r-ryantm.': pull_request.user.login === 'r-ryantm',
```
I personally feel like "authoring" is a bit more creative work. But r-ryantm is not creative, as long as we don't let it do AI-updates ;)

Could we use "created" for committers, too? Maybe. I just didn't want to use authored for r-ryantm.

MattSturgeon replied:
I guess for me this highlighted a connotation with `git` authors & committers; the person who authored the commits in a PR may not be the same person who created (or _opened_) the PR.

In fact, for a single commit PR, you could feasibly have a different opener/creator, committer, and author. E.g. this happens when Person A force-pushes a rebase to a backport PR opened by nixpkgs-ci, backporting a commit authored by Person B.

wolfgangwalther commented on ci/README.md:
```diff
@@ -43,7 +43,7 @@ These issues effectively list PRs the merge bot has interacted with.
 
 - The PR targets `master`, `staging`, or `staging-next`.
-- The PR only touches files located under `pkgs/by-name/*`.
+- The PR only touches packages located under `pkgs/by-name/*`.
```
Uhm, but "all files are inside `pkgs/by-name`" is not correct. When you edit `pkgs/by-name/README.md`, this would be included in your definition (the one with files), but *not* when I say "packages", because the README is not a package.

MattSturgeon replied:
That's a good edge-case to consider. Other edge-cases might include miscellaneous files _related_ to a package but not directly _part of_ the package.

`pkgs/by-name/README.md` would satisfy the `allByName` constraint, though:

```js
  const allByName = files.every(({ filename }) => 
     filename.startsWith('pkgs/by-name/'), 
   )
```

Which is one reason I suggested the "all changed files are in by-name" wording.

wolfgangwalther replied:
> That's a good edge-case to consider. Other edge-cases might include miscellaneous files _related_ to a package but not directly _part of_ the package.

Hm. I consider the enforced rules of by-name to mean that "everything related to a package is by definition inside the respective by-name folder".

> `pkgs/by-name/README.md` would satisfy the `allByName` constraint, though:

Uh, right, I see what you mean. Will be fixed in #458020.
